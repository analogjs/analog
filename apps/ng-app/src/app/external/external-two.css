h1 {
  color: darkblue;
}
