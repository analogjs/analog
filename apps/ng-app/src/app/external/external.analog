<script lang="ts">
import { signal } from '@angular/core';

defineMetadata({
  templateUrl: './external.html',
  styleUrls: ['./external.css', './external-two.css'],
})

const value = signal('');
</script>
