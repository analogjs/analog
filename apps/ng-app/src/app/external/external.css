p {
  color: tomato;
  background: #eee;
}
