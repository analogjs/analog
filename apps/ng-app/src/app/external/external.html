<div>
  <h1>I am in external template!!!</h1>
  <input [value]="value()" (input)="value.set($any($event.target).value)" />
  <p>Value: {{ value() }}</p>
</div>
