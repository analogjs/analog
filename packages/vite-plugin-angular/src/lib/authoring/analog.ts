import {
  ArrowFunction,
  ClassDeclaration,
  ConstructorDeclaration,
  FunctionDeclaration,
  FunctionExpression,
  ImportAttributeStructure,
  ImportDeclaration,
  ImportSpecifierStructure,
  Node,
  ObjectLiteralExpression,
  OptionalKind,
  Project,
  PropertyAssignment,
  PropertyDeclarationStructure,
  Scope,
  SourceFile,
  StructureKind,
  SyntaxKind,
  VariableDeclarationKind,
} from 'ts-morph';
import {
  HOOKS_MAP,
  INVALID_METADATA_PROPERTIES,
  ON_DESTROY,
  ON_INIT,
  REQUIRED_SIGNALS_MAP,
  SCRIPT_TAG_REGEX,
  SIGNALS_MAP,
  STYLE_TAG_REGEX,
  TEMPLATE_TAG_REGEX,
} from './constants.js';

import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);

export function compileAnalogFile(
  filePath: string,
  fileContent: string,
  shouldFormat = false
) {
  const componentName = filePath.split('/').pop()?.split('.')[0];
  if (!componentName) {
    throw new Error(`[Analog] Missing component name ${filePath}`);
  }

  const [componentFileName, className] = [
    toFileName(componentName),
    toClassName(componentName),
  ];

  const isMarkdown = fileContent.includes('lang="md"');

  const [scriptContent, styleContent] = [
    SCRIPT_TAG_REGEX.exec(fileContent)?.pop()?.trim() || '',
    STYLE_TAG_REGEX.exec(fileContent)?.pop()?.trim() || '',
  ];

  const templateContentMatches = TEMPLATE_TAG_REGEX.exec(fileContent) ?? [];
  const templateAttributeString =
    templateContentMatches[2]?.trim().replace(/\n/g, '') ?? '';

  let templateContent = '';

  if (!isMarkdown) {
    templateContent = templateContentMatches[3]?.trim() || '';
  }

  let ngType: 'Component' | 'Directive';
  if (templateContent || isMarkdown) {
    ngType = 'Component';
  } else if (scriptContent && !templateContent) {
    ngType = scriptContent.includes('templateUrl') ? 'Component' : 'Directive';
  } else {
    throw new Error(`[Analog] Cannot determine entity type ${filePath}`);
  }

  const entityName = `${className}Analog${ngType}`;
  const componentMetadata = (() => {
    if (ngType === 'Component') {
      const items = ['changeDetection: ChangeDetectionStrategy.OnPush'];

      if (isMarkdown) {
        items.push(
          `templateUrl: \`virtual-analog:${filePath.replace('.ts', '')}\``
        );
      } else if (templateContent) {
        items.push(`template: \`${templateContent}\``);
      }

      if (styleContent) {
        items.push(`styles: \`${styleContent.replaceAll('\n', '')}\``);
      }

      if (templateAttributeString) {
        const attributes = parseAttributes(templateAttributeString);
        items.push(`host: ${JSON.stringify(attributes)}`);
      }

      return items.join(',\n  ');
    }
    return '';
  })();

  const source = `
import { ${ngType}${
    ngType === 'Component' ? ', ChangeDetectionStrategy' : ''
  } } from '@angular/core';

@${ngType}({
  standalone: true,
  selector: '${componentFileName},${className}',
  ${componentMetadata}
})
export default class ${entityName} {
  constructor() {}
}`;

  // the `.analog` file
  if (scriptContent) {
    const project = new Project({ useInMemoryFileSystem: true });
    return processAnalogScript(
      filePath,
      project.createSourceFile(filePath, scriptContent),
      project.createSourceFile(`${filePath}.virtual.ts`, source),
      ngType,
      entityName,
      shouldFormat
    );
  }

  return source;
}

function processAnalogScript(
  fileName: string,
  ngSourceFile: SourceFile,
  targetSourceFile: SourceFile,
  ngType: 'Component' | 'Directive',
  entityName: string,
  isProd?: boolean
) {
  const targetClass = targetSourceFile.getClass(
    (classDeclaration) => classDeclaration.getName() === entityName
  );

  if (!targetClass) {
    throw new Error(`[Analog] Missing class ${fileName}`);
  }

  const targetMetadata = targetClass.getDecorator(ngType);

  if (!targetMetadata) {
    throw new Error(`[Analog] Missing metadata ${fileName}`);
  }

  const targetMetadataArguments =
    targetMetadata.getArguments()[0] as ObjectLiteralExpression;

  if (!Node.isObjectLiteralExpression(targetMetadataArguments)) {
    throw new Error(`[Analog] invalid metadata arguments ${fileName}`);
  }

  const targetConstructor = targetClass.getConstructors()[0],
    targetConstructorBody = targetConstructor.getBody();

  if (!Node.isBlock(targetConstructorBody)) {
    throw new Error(`[Analog] invalid constructor body ${fileName}`);
  }

  const importAttributes: { [key: string]: Array<string> } = {
    imports: [],
    viewProviders: [],
    providers: [],
    exposes: [],
  };

  const gettersSetters: Array<{ propertyName: string; isFunction: boolean }> =
      [],
    outputs: Array<string> = [],
    sourceSyntaxList = ngSourceFile.getChildren()[0]; // SyntaxList

  if (!Node.isSyntaxList(sourceSyntaxList)) {
    throw new Error(`[Analog] invalid source syntax list ${fileName}`);
  }

  for (const node of sourceSyntaxList.getChildren()) {
    if (Node.isImportDeclaration(node)) {
      let structure = node.getStructure();

      if (
        !structure.namedImports?.length &&
        !structure.defaultImport &&
        structure.moduleSpecifier.endsWith('.analog')
      ) {
        const generatedName = structure.moduleSpecifier.replace(
          /[^a-zA-Z]/g,
          ''
        );
        (node as ImportDeclaration).setDefaultImport(generatedName);
        structure = node.getStructure();
      }

      const attributes = structure.attributes;
      const passThroughAttributes: OptionalKind<ImportAttributeStructure>[] =
        [];
      let foundAttribute = '';

      for (const attribute of attributes || []) {
        if (attribute.name === 'analog') {
          const value = attribute.value.replaceAll("'", '');
          if (!(value in importAttributes)) {
            throw new Error(
              `[Analog] Invalid Analog import attribute ${value} in ${fileName}`
            );
          }
          foundAttribute = value;
          continue;
        }

        passThroughAttributes.push(attribute);
      }

      if (foundAttribute) {
        const { defaultImport, namedImports } = structure;
        if (defaultImport) {
          importAttributes[foundAttribute].push(defaultImport);
        }

        if (namedImports && Array.isArray(namedImports)) {
          const namedImportStructures = namedImports.filter(
            (
              namedImport
            ): namedImport is OptionalKind<ImportSpecifierStructure> =>
              typeof namedImport === 'object'
          );
          const importNames = namedImportStructures.map(
            (namedImport) => namedImport.alias ?? namedImport.name
          );
          importAttributes[foundAttribute].push(...importNames);
        }
      }

      // copy the import to the target `.analog.ts` file
      targetSourceFile.addImportDeclaration({
        ...structure,
        attributes: passThroughAttributes.length
          ? passThroughAttributes
          : undefined,
      });
      continue;
    }

    const nodeFullText = node.getFullText();

    if (Node.isExportable(node) && node.hasExportKeyword()) {
      targetSourceFile.addStatements(nodeFullText);
      continue;
    }

    if (Node.isVariableStatement(node)) {
      // NOTE: we do not support multiple declarations (i.e: const a, b, c)
      const declaration = node.getDeclarations()[0],
        isLet = node.getDeclarationKind() === VariableDeclarationKind.Let;

      const name = declaration.getName(),
        initializer = declaration.getInitializer();

      if (!initializer && isLet) {
        // transfer the whole line `let variable;` over
        addVariableToConstructor(targetConstructor, '', name, 'let');
        // populate getters array for Object.defineProperties
        gettersSetters.push({ propertyName: name, isFunction: false });

        continue;
      }

      if (initializer) {
        // with initializer
        const nameNode = declaration.getNameNode();

        // if destructured.
        // TODO: we don't have a good abstraction for handling destructured variables yet.
        if (
          Node.isArrayBindingPattern(nameNode) ||
          Node.isObjectBindingPattern(nameNode)
        ) {
          targetConstructor.addStatements(nodeFullText);

          const bindingElements = nameNode
            .getDescendantsOfKind(SyntaxKind.BindingElement)
            .map((bindingElement) => bindingElement.getName());

          if (isLet) {
            gettersSetters.push(
              ...bindingElements.map((propertyName) => ({
                propertyName,
                isFunction: false,
              }))
            );
            continue;
          }

          for (const bindingElement of bindingElements) {
            targetClass.addProperty({
              name: bindingElement,
              kind: StructureKind.Property,
              scope: Scope.Protected,
            });
            targetConstructor.addStatements(
              `this.${bindingElement} = ${bindingElement};`
            );
          }
          continue;
        }

        const isFunctionInitializer = isFunction(initializer);

        if (!isLet) {
          const ioStructure = getIOStructure(initializer);

          if (ioStructure) {
            // outputs
            if (ioStructure.decorators) {
              // track output name
              outputs.push(name);
            }

            // track output name
            targetClass.addProperty({
              ...ioStructure,
              decorators: undefined,
              name,
              scope: Scope.Protected,
            });

            // assign constructor variable
            targetConstructor.addStatements(`const ${name} = this.${name}`);

            continue;
          }

          /**
           * normal property
           * const variable = initializer;
           * We'll create a class property with the same variable name
           */
          addVariableToConstructor(
            targetConstructor,
            initializer.getText(),
            name,
            'const',
            true
          );
          continue;
        }

        /**
         * let variable = initializer;
         * We will NOT create a class property with the name because we will add it to the getters
         */
        addVariableToConstructor(
          targetConstructor,
          initializer.getText(),
          name,
          'let'
        );
        gettersSetters.push({
          propertyName: name,
          isFunction: isFunctionInitializer,
        });

        continue;
      }

      continue;
    }

    if (Node.isFunctionDeclaration(node)) {
      const functionName = node.getName();
      if (functionName) {
        targetConstructor.addStatements([
          // bring the function over
          nodeFullText,
          // assign class property
          `this.${functionName} = ${functionName}.bind(this);`,
        ]);

        continue;
      }

      continue;
    }

    if (Node.isExpressionStatement(node)) {
      const expression = node.getExpression();

      if (Node.isCallExpression(expression)) {
        // hooks, effects, basically Function calls
        const functionName = expression.getExpression().getText();
        if (functionName === 'defineMetadata') {
          const metadata =
            expression.getArguments()[0] as ObjectLiteralExpression;
          const metadataProperties = metadata
            .getPropertiesWithComments()
            .filter(
              (property): property is PropertyAssignment =>
                Node.isPropertyAssignment(property) &&
                !INVALID_METADATA_PROPERTIES.includes(property.getName())
            );

          if (metadataProperties.length === 0) continue;

          processMetadata(
            metadataProperties,
            targetMetadataArguments,
            targetClass
          );
          continue;
        }

        if (functionName === ON_INIT || functionName === ON_DESTROY) {
          const initFunction = expression.getArguments()[0];
          if (isFunction(initFunction)) {
            // add the function to constructor
            targetConstructor.addStatements(
              `this.${functionName} = ${initFunction.getText()}`
            );

            // add life-cycle method to class
            targetClass.addMethod({
              name: HOOKS_MAP[functionName],
              statements: `this.${functionName}();`,
            });
            continue;
          }

          continue;
        }

        // just add the entire node to the constructor. i.e: effect()
        targetConstructor.addStatements(nodeFullText);
      }
    }
  }

  if (ngType === 'Component') {
    if (importAttributes['viewProviders'].length) {
      processArrayLiteralMetadata(
        targetMetadataArguments,
        'viewProviders',
        importAttributes['viewProviders']
      );
    }

    if (importAttributes['imports'].length) {
      processArrayLiteralMetadata(
        targetMetadataArguments,
        'imports',
        importAttributes['imports']
      );
    }

    if (importAttributes['exposes'].length) {
      const exposes = importAttributes['exposes'].map((item) => ({
        name: item.trim(),
        initializer: item.trim(),
        scope: Scope.Protected,
      }));

      targetClass.addProperties(exposes);
    }
  }

  if (importAttributes['providers'].length) {
    processArrayLiteralMetadata(
      targetMetadataArguments,
      'providers',
      importAttributes['providers']
    );
  }

  if (outputs.length) {
    processArrayLiteralMetadata(
      targetMetadataArguments,
      'outputs',
      outputs.map((output) => `'${output}'`)
    );
  }

  if (gettersSetters.length) {
    targetConstructor.addStatements(`
Object.defineProperties(this, {
${gettersSetters
  .map(
    ({ isFunction, propertyName }) =>
      `${propertyName}:{get(){return ${
        !isFunction ? `${propertyName}` : `${propertyName}.bind(this)`
      };},set(v){${propertyName}=v;}},`
  )
  .join('\n')}
});`);
  }

  if (!isProd) {
    // PROD probably does not need this
    targetSourceFile.formatText({ ensureNewLineAtEndOfFile: true });
  }

  return targetSourceFile.getText();
}

function processArrayLiteralMetadata(
  targetMetadataArguments: ObjectLiteralExpression,
  metadataName: string,
  items: string[]
) {
  let metadata = targetMetadataArguments.getProperty(metadataName);

  if (!metadata) {
    metadata = targetMetadataArguments.addPropertyAssignment({
      name: metadataName,
      initializer: '[]',
    });
  }

  const initializer =
    Node.isPropertyAssignment(metadata) && metadata.getInitializer();

  if (initializer && Node.isArrayLiteralExpression(initializer)) {
    initializer.addElements(items);
  }
}

function processMetadata(
  metadataProperties: PropertyAssignment[],
  targetMetadataArguments: ObjectLiteralExpression,
  targetClass: ClassDeclaration
) {
  metadataProperties.forEach((property) => {
    const propertyInitializer = property.getInitializer();
    if (propertyInitializer) {
      const propertyName = property.getName(),
        propertyInitializerText = propertyInitializer.getText();

      if (propertyName === 'selector') {
        // remove the existing selector
        targetMetadataArguments.getProperty('selector')?.remove();
        // add the new selector
        targetMetadataArguments.addPropertyAssignment({
          name: 'selector',
          initializer: propertyInitializerText,
        });
      } else if (propertyName === 'exposes') {
        // for exposes we're going to add the property to the class so they are accessible on the template
        // parse the initializer to get the item in the exposes array
        const exposes = propertyInitializerText
          .replace(/[[\]]/g, '')
          .split(',')
          .map((item) => ({
            name: item.trim(),
            initializer: item.trim(),
            scope: Scope.Protected,
          }));

        targetClass.addProperties(exposes);
      } else {
        targetMetadataArguments.addPropertyAssignment({
          name: propertyName,
          initializer: propertyInitializerText,
        });
      }
    }
  });
}

/**
 * const variable = initializer;
 * |
 * v
 * constructor() {
 *  const variable = (this.variable = initializer);
 * }
 *
 */
function addVariableToConstructor(
  targetConstructor: ConstructorDeclaration,
  initializerText: string,
  variableName: string,
  kind: 'let' | 'const',
  withClassProperty = false
) {
  let statement = `${kind} ${variableName}`;

  if (initializerText) {
    statement += `= ${initializerText}`;

    if (withClassProperty) {
      statement = `${kind} ${variableName} = (this.${variableName} = ${initializerText})`;
    }
  }

  targetConstructor.addStatements(statement + ';');
}

function isFunction(
  initializer: Node
): initializer is ArrowFunction | FunctionDeclaration | FunctionExpression {
  return (
    Node.isArrowFunction(initializer) ||
    Node.isFunctionDeclaration(initializer) ||
    Node.isFunctionExpression(initializer)
  );
}

function getIOStructure(
  initializer: Node
): Omit<OptionalKind<PropertyDeclarationStructure>, 'name'> | null {
  const callableExpression =
    (Node.isCallExpression(initializer) || Node.isNewExpression(initializer)) &&
    initializer;

  if (!callableExpression) return null;

  const expression = callableExpression.getExpression(),
    initializerText = callableExpression.getText();

  if (initializerText.includes('new EventEmitter')) {
    return {
      initializer: initializerText,
      decorators: [{ name: 'Output', arguments: [] }],
    };
  }

  if (
    (Node.isPropertyAccessExpression(expression) &&
      REQUIRED_SIGNALS_MAP[expression.getText()]) ||
    (Node.isIdentifier(expression) && SIGNALS_MAP[expression.getText()])
  ) {
    return { initializer: initializer.getText() };
  }

  return null;
}

/**
 * Hyphenated to UpperCamelCase
 */
function toClassName(str: string) {
  return toCapitalCase(toPropertyName(str));
}
/**
 * Hyphenated to lowerCamelCase
 */
function toPropertyName(str: string) {
  return str
    .replace(/([^a-zA-Z0-9])+(.)?/g, (_, __, chr) =>
      chr ? chr.toUpperCase() : ''
    )
    .replace(/[^a-zA-Z\d]/g, '')
    .replace(/^([A-Z])/, (m) => m.toLowerCase())
    .replace(/^\d+/, '');
}

/**
 * Upper camelCase to lowercase, hyphenated
 */
function toFileName(str: string) {
  return str
    .replace(/([a-z\d])([A-Z])/g, '$1_$2')
    .toLowerCase()
    .replace(/(?!^[_])[ _]/g, '-')
    .replace(/^\d+-?/, '');
}
/**
 * Capitalizes the first letter of a string
 */
function toCapitalCase(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

/**
 * Parses the attribute string and returns a key-value pair
 *
 * @param attributeString
 */
function parseAttributes(attributeString: string): Record<string, string> {
  const attributes: Record<string, string> = {};
  const pattern = /([^\s"'=]+)=?(?:(["'])(.*?\2)|(.+?))(?=\s|$)/g; // Regex to capture key-value pairs

  let match: RegExpExecArray | null;
  while ((match = pattern.exec(attributeString)) !== null) {
    const [, key, , quotedValue, unquotedValue] = match;
    let value = quotedValue || unquotedValue || ''; // Use quoted or unquoted value
    if (value.endsWith('"')) {
      value = value.slice(0, -1);
    }
    attributes[key.trim()] = value.trim();
  }

  return attributes;
}
