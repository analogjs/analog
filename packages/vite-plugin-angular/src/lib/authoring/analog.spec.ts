import { compileAnalogFile } from './analog';

const COMPONENT_CONTENT = `
<script lang="ts">
import { signal, input, ViewChild, afterNextRender, ElementRef, viewChild, viewChildren, contentChild, contentChildren, output, model } from '@angular/core';
import External from './external.analog' with { analog: 'imports' };
import { ExternalService } from './external' with { analog: 'providers' };
import { ExternalEnum } from './external.model' with { analog: 'exposes' };
import './noname.analog' with { analog: 'imports' };

defineMetadata({
  exposes: [Math],
  queries: {
    divElement: new ViewChild('divElement')
  }
});

export const routeMeta = {
  title: 'My page',
  canActivate: [() => true],
}

export interface MyInterface {
  title: string
}

export type MyType = string;

export enum Direction {
  Up,
  Down,
  Left,
  Right,
}

export function myFunc(){
  console.log('hello');
}

let divElement: ElementRef<HTMLDivElement>;
let test: string;

setTimeout(() => {
  test = 'test';
}, 1000)

const counter = signal(0);
const [a, b, , c = 4] = [1, 2, 3];

const inputWithDefault = input(""); // InputSignal<string, string>
const inputWithoutDefault = input<string>(); // InputSignal<string | undefined, string | undefined>
const inputWithAlias = input("", { alias: "theAlias" }); // InputSignal<string, string>
const inputWithoutDefaultWithAlias = input<string | undefined>(undefined, {
    alias: "theAlias",
  }); // InputSignal<string | undefined, string | undefined>
const inputWithTransform = input<unknown, boolean>("", {
    transform: booleanAttribute,
  }); // InputSignal<unknown, boolean>
const requiredInput = input.required<string>(); // InputSignal<string, string>
const requiredInputWithTransform = input.required<unknown, number>({
    transform: (value) => numberAttribute(value, 10),
  });
const output = new EventEmitter();
const newOutput = output();
const outputWithType = new EventEmitter<string>();

const myModel = model();

const viewChildEl = viewChild<HTMLDivElement>('divElement');
const viewChildRequiredEl = viewChild.required<HTMLDivElement>('divElement');

const viewChildrenEl = viewChildren<HTMLDivElement[]>('divElement');

const contentChildEl = contentChild<ElementRef>('divElement');
const contentChildRequiredEl = contentChild.required<ElementRef>('divElement');

const contentChildrenEl = contentChildren<HTMLDivElement[]>('divElement');

afterNextRender(() => {
  console.log('the div', divElement);
})

</script>

<template class="flex flex-col" [class.binding]="counter()" (click)="handleClick($event)">
  <div #divElement>Component</div>
  <p>{{ counter() }}</p>
  <p>{ a }</p>
  <p>{ b }</p>
  <p>{ c }</p>
  <p>{{ test }}</p>
</template>

<style>
  div {
    color: red;
  }

  p {
    color: blue;
  }
</style>
`;

const DIRECTIVE_CONTENT = `
<script lang="ts">
import { inject, ElementRef,afterNextRender } from '@angular/core';

defineMetadata({
  selector: 'input[directive]',
})

const elRef = inject(ElementRef);

afterNextRender(() => {
  elRef.nativeElement.focus();
});

onInit(() => {
  console.log('init code');
});

effect(() => {
  console.log('just some effect');
});
</script>
`;

const MARKDOWN_CONTENT = `
<template
  lang="md"
  class="flex flex-col"
  [style]="{ color: 'red' }"
 >
  # Hello World
</template>
`;

describe('authoring ng file', () => {
  it('should process component as an analog file', () => {
    const source = compileAnalogFile('virtual.analog.ts', COMPONENT_CONTENT);
    expect(source).toContain('Component');
    expect(source).toMatchSnapshot();
  });

  it('should process directive as an analog file', () => {
    const source = compileAnalogFile('virtual.analog.ts', DIRECTIVE_CONTENT);
    expect(source).toContain('Directive');
    expect(source).toMatchSnapshot();
  });

  it('should process component with markdown as analog file with a virtual templateUrl', () => {
    const source = compileAnalogFile('virtual.analog.ts', MARKDOWN_CONTENT);
    expect(source).toContain('Component');
    expect(source).toMatchSnapshot();
  });
});
