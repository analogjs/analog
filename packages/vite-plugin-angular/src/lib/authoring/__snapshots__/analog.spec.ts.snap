// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`authoring ng file > should process component as an analog file 1`] = `
"import { Component, ChangeDetectionStrategy } from '@angular/core';
import { signal, input, ViewChild, afterNextRender, ElementRef, viewChild, viewChildren, contentChild, contentChildren, output, model } from "@angular/core";
import External from "./external.analog";
import { ExternalService } from "./external";
import { ExternalEnum } from "./external.model";
import nonameanalog from "./noname.analog";

@Component({
    standalone: true,
    selector: 'virtual,Virtual',
    changeDetection: ChangeDetectionStrategy.OnPush,
    template: \`<div #divElement>Component</div>
  <p>{{ counter() }}</p>
  <p>{ a }</p>
  <p>{ b }</p>
  <p>{ c }</p>
  <p>{{ test }}</p>\`,
    styles: \`div {    color: red;  }  p {    color: blue;  }\`,
    host: { "class": "flex flex-col", "[class.binding]": "counter()", "(click)": "handleClick($event)" },
    queries: {
        divElement: new ViewChild('divElement')
    },
    imports: [External, nonameanalog],
    providers: [ExternalService],
    outputs: ['output', 'outputWithType']
})
export default class VirtualAnalogComponent {
    constructor() {
        let divElement;
        let test;


        setTimeout(() => {
            test = 'test';
        }, 1000)
        const counter = (this.counter = signal(0));

        const [a, b, , c = 4] = [1, 2, 3];
        this.a = a;
        this.b = b;
        this.c = c;
        const inputWithDefault = this.inputWithDefault
        const inputWithoutDefault = this.inputWithoutDefault
        const inputWithAlias = this.inputWithAlias
        const inputWithoutDefaultWithAlias = this.inputWithoutDefaultWithAlias
        const inputWithTransform = this.inputWithTransform
        const requiredInput = this.requiredInput
        const requiredInputWithTransform = this.requiredInputWithTransform
        const output = this.output
        const newOutput = this.newOutput
        const outputWithType = this.outputWithType
        const myModel = this.myModel
        const viewChildEl = this.viewChildEl
        const viewChildRequiredEl = this.viewChildRequiredEl
        const viewChildrenEl = this.viewChildrenEl
        const contentChildEl = this.contentChildEl
        const contentChildRequiredEl = this.contentChildRequiredEl
        const contentChildrenEl = this.contentChildrenEl


        afterNextRender(() => {
            console.log('the div', divElement);
        })

        Object.defineProperties(this, {
            divElement: { get() { return divElement; }, set(v) { divElement = v; } },
            test: { get() { return test; }, set(v) { test = v; } },
        });
    }

    protected Math = Math;
    protected a;
    protected b;
    protected c;
    protected inputWithDefault = input("");
    protected inputWithoutDefault = input<string>();
    protected inputWithAlias = input("", { alias: "theAlias" });
    protected inputWithoutDefaultWithAlias = input<string | undefined>(undefined, {
        alias: "theAlias",
    });
    protected inputWithTransform = input<unknown, boolean>("", {
        transform: booleanAttribute,
    });
    protected requiredInput = input.required<string>();
    protected requiredInputWithTransform = input.required<unknown, number>({
        transform: (value) => numberAttribute(value, 10),
    });
    protected output = new EventEmitter();
    protected newOutput = output();
    protected outputWithType = new EventEmitter<string>();
    protected myModel = model();
    protected viewChildEl = viewChild<HTMLDivElement>('divElement');
    protected viewChildRequiredEl = viewChild.required<HTMLDivElement>('divElement');
    protected viewChildrenEl = viewChildren<HTMLDivElement[]>('divElement');
    protected contentChildEl = contentChild<ElementRef>('divElement');
    protected contentChildRequiredEl = contentChild.required<ElementRef>('divElement');
    protected contentChildrenEl = contentChildren<HTMLDivElement[]>('divElement');
    protected ExternalEnum = ExternalEnum;
}


export const routeMeta = {
    title: 'My page',
    canActivate: [() => true],
}


export interface MyInterface {
    title: string
}


export type MyType = string;


export enum Direction {
    Up,
    Down,
    Left,
    Right,
}


export function myFunc() {
    console.log('hello');
}
"
`;

exports[`authoring ng file > should process component with markdown as analog file with a virtual templateUrl 1`] = `
"
import { Component, ChangeDetectionStrategy } from '@angular/core';

@Component({
  standalone: true,
  selector: 'virtual,Virtual',
  changeDetection: ChangeDetectionStrategy.OnPush,
  templateUrl: \`virtual-analog:virtual.analog\`,
  host: {"lang":"md","class":"flex flex-col","[style]":"{ color: 'red' }"}
})
export default class VirtualAnalogComponent {
  constructor() {}
}"
`;

exports[`authoring ng file > should process directive as an analog file 1`] = `
"import { Directive } from '@angular/core';
import { inject, ElementRef, afterNextRender } from "@angular/core";

@Directive({
    standalone: true,
    selector: 'input[directive]'
})
export default class VirtualAnalogDirective {
    constructor() {
        const elRef = (this.elRef = inject(ElementRef));


        afterNextRender(() => {
            elRef.nativeElement.focus();
        });
        this.onInit = () => {
            console.log('init code');
        }


        effect(() => {
            console.log('just some effect');
        });
    }

    ngOnInit() {
        this.onInit();
    }
}
"
`;
